engine/runtime-instrument-common/src/main/java/org/enso/interpreter/service/ExecutionService.java

```diff
@@ -76,6 +77,7 @@ public final class ExecutionService {
   private final CallRootNode call = new CallRootNode();
   private final InvokeMemberRootNode invoke = new InvokeMemberRootNode();
   private final Timer timer;
+  private RuntimeCache lastExecCache;
```

Comment:
I've been pretty [unpleasantly surprised](https://discord.com/channels/@me/948519891961544714/1217372688704606290) there is more than a single `RuntimeCache`. @4e6 wrote:

> We have one RuntimeCache per execution context, plus a separate cache per visualization. The visualization cache is supposed to contain intermediate computations of the visualization function. Because the execution of visualization expression is a separate program execution, it makes sense to have a separate cache for it. Plus it makes the cache invalidation simple.

I am not convinced. I'd really love to [access all the available cached values](https://github.com/enso-org/enso/pull/9410/files#r1524440594)...

Reply:
We discuseed the ability to use the hierarchical cache, and it sounds like what you need. 

> We have one RuntimeCache per execution context, plus a separate cache per visualization.

The visualization cache should have read access to the `RuntimeCache` of the execution context. This way you will get access to the computed values.

Reply:
I managed to get rid of the `lastExecCache` field in 4c62581 - @4e6, please verify that's correct. The next step is to get rid of the [ThreadLocal](https://github.com/enso-org/enso/pull/9410#discussion_r1524444751) - I am not sure I need the _hierarchical cache_ for that as I am still

> [unpleasantly surprised](https://discord.com/channels/@me/948519891961544714/1217372688704606290) there is more than a single RuntimeCache

I think there should be one `RuntimeCache` per the `IdExecutionInstrument`!